djpsa/sync/sync.py

```
@@ -73,6 +73,49 @@ def __init__(self):
         self.synced_ids = set()
 
 
+class SyncRelatedMixin:
+    """
+    Mixin for synchronizers that need to sync related objects.
+    """
+
+    def sync_related(self, instance, sync_classes=None):
+        """
+        Sync related objects for the given instance.
+        """
+        sync_classes = sync_classes + self.get_related_synchronizers(instance) \
+            if sync_classes else self.get_related_synchronizers(instance)
+
+
+        for sync_class in sync_classes:
+            self._relation_sync(*sync_class)
+
+    def get_related_synchronizers(self, instance):
+        """
+        Return a list of related synchronizers.
+        """
+        raise NotImplementedError
```

I don't think a mixin should raise this exception. Mixins are for adding useful behaviour that's common among a few classes, not for _requiring_ a child class to do something. Perhaps the default implementation could just be a no-op, but this might call for some re-thinking the overall design because it seems off.

Does this need to be a mixin? Is syncing related things not a general enough need for synchronizes that this could just be part of the main parent?

---

So, I've thought a lot about this, and I think it should be a separate class. It goes against two good OO design principles, don't give classes interfaces they don't need, and open for extension, closed for modification. Obviously everyone breaks these often but my metric is "how far are we breaking them". There are way more synchronizers that will never use this than will use this, and adding things to the base sync class, or api class, etc is how we got blobs in djconnectwise and djautotask, so I'd like to avoid that with our clean slate.

I definitely agree that this mixin shouldn't be raising NotImplementedError, and I think the big issue is that I named it mixin out of habit, and made it look like a mixin, when what I was writing was an abstract class. So instead I think it would be better to use Python's abstract class framework `abc`. Now I know this isn't a pattern we have used in the past, but we are starting with a blank slate, I don't want to avoid using tools in the toolbox just because we haven't had them. This way, we keep the benefits of extension over modification, and only Synchronizers that need it can implement it. Its totally separate from the sync class, so developers know they can make changes to it without breaking the synchronizer, and new developers in the future are less likely to make changes that will couple the synchronizer to the sync related functionality (coupling sync related to synchronizer is fine of course, because it wouldn't make sense without it). By using `abc`, instead of raising a `NotImplementedError`, it just passes and we decorate the function with `@abstractmethod`. So the IDE will tell the developer and the reviewer that the class is incomplete without the specific definitions, a nice feature that python mostly lacks.

Now, if I were to do this over again I think I would go with composition over inheritance. Since all django REALLY favours inheritance over composition I still lean on it too much. It might be good to make a new class "SyncRelatedBlahBlah" that either takes in data or has a child class for different record types (Ticket, Opportunity, Project, etc), and then all the differences are encapsulated in there. I think if you really don't want to use the `abc` package that is what I will do, because I am against just putting things in the base sync class, especially if the majority of synchronizers wont use it.